wdl-analysis/src/types.rs

```diff
@@ -0,0 +1,2060 @@
+//! Representation of the WDL type system.
+
+use std::fmt;
+
+use id_arena::Arena;
+use id_arena::ArenaBehavior;
+use id_arena::DefaultArenaBehavior;
+use id_arena::Id;
+use indexmap::IndexMap;
+
+use crate::STDLIB;
+
+/// A trait implemented on types that may be optional.
+pub trait Optional: Copy {
+    /// Determines if the type is optional.
+    fn is_optional(&self) -> bool;
+
+    /// Makes the type optional if it isn't already optional.
+    fn optional(&self) -> Self;
+
+    /// Makes the type required if it isn't already required.
+    fn require(&self) -> Self;
+}
+
+/// A trait implemented on types that are coercible to other types.
+pub trait Coercible {
+    /// Determines if the type is coercible to the target type.
+    fn is_coercible_to(&self, types: &Types, target: &Self) -> bool;
```

Comment:
I wonder if this would be better served as an `Option` that returns the coerced type? If nothing else, to save needing to check and then doing the actual coercing.